### PR TITLE
Create untitled notebook names in core code rather than in the notebook extension

### LINKED ISSUE #21073: New notebook sends me to existing unsaved notebook-x
Issue Type: <b>Bug</b>

This happens when I reopen azure data studio and I have several unsaved notebooks (Notebook-0, 1, 2,...)
Then I go to a server and right click create new notebook, and it sends me to the unsaved already existing Notebook-0, if I clic create new nootebook again it sends me to NB-1 (that already exists and was unsaved), and I have to do this several times (nb 2 then nb 3...)until I finally get a really new notebook,

Azure Data Studio version: azuredatastudio 1.39.1 (7553f799e175f471b7590302dd65c997b838b29b, 2022-08-30T01:11:13.047Z)
OS version: Windows_NT x64 10.0.22621
Restricted Mode: No

<details><summary>Extensions (6)</summary>

Extension|Author (truncated)|Version
---|---|---
admin-pack|Mic|0.0.2
agent|Mic|0.49.0
dacpac|Mic|1.10.1
import|Mic|1.5.0
kusto|Mic|0.5.4
profiler|Mic|0.11.0

</details>

### PR DESCRIPTION
**Pull request by corivera**
This PR fixes #21073

Currently when creating an untitled notebook, we create the untitled tab name (e.g. "Notebook-1") from the notebook extension, but the azdata.nb.notebookDocuments field we use to check for duplicate title names only gets updated when a notebook tab is clicked on. So, if you open ADS with 10 notebook tabs already open, but only click on the first one, then the notebookDocuments field will only contain that first notebook. When opening a new untitled notebook in that situation, we end up reopening a tab with the same name, rather than creating a notebook number higher than all the other existing ones. Like if we have a tab "Notebook-2" already open, but hasn't been clicked on, if we create a second untitled notebook it will reopen Notebook-2 instead of creating a new one.

This change moves untitled notebook name generation to the core code, so that we check the UntitledEditorService directly like opening a query editor does, instead of using the notebookDocuments field in the extension API. For this I re-used our existing azdata.nb.showNotebookDocument function, and added checks for duplicate notebook names when trying to open an untitled notebook.

### CODE REVIEW

File: src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts [hunk 1]
```diff
@@ -343,15 +345,47 @@ export class MainThreadNotebookDocumentsAndEditors extends Disposable implements
 		}
 	}
 
+	/**
+	 * Attempts to create a new notebook with the specified provider and contents. Used for VS Code extension compatibility.
```
Comment (corivera) [inline src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts:349]: Added some comments here for the VS Code compatibility stuff to reduce confusion with our own notebook creation methods.

File: extensions/notebook/src/test/common/notebookUtils.test.ts [hunk 2]
```diff
@@ -50,12 +50,12 @@ describe('notebookUtils Tests', function (): void {
 			should(azdata.nb.notebookDocuments.length).equal(0, 'There should be not any open Notebook documents');
 			await notebookUtils.newNotebook(undefined);
 			should(azdata.nb.notebookDocuments.length).equal(1, 'There should be exactly 1 open Notebook document');
-			should(azdata.nb.notebookDocuments[0].fileName).equal('Notebook-0', 'The first Untitled Notebook should have an index of 0');
+			should(azdata.nb.notebookDocuments[0].fileName).equal('Notebook-1', 'The first Untitled Notebook should have an index of 1');
```
Comment (corivera) [inline extensions/notebook/src/test/common/notebookUtils.test.ts:53]: Untitled Text and Query document names start counting up from 1, so I updated notebooks to have the same naming behavior.

File: src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts [hunk 3]
```diff
@@ -343,15 +345,47 @@ export class MainThreadNotebookDocumentsAndEditors extends Disposable implements
...
+			}
+		}
 		return Promise.resolve(this.doOpenEditor(resource, options));
 	}
 
+	private untitledEditorTitleExists(filePath: string): boolean {
+		return !!this._untitledEditorService.get(URI.from({ scheme: Schemas.untitled, path: filePath }));
+	}
+
+	private createPrefixedNotebookFilePath(prefix?: string): string {
```
Comment (corivera) [inline src/sql/workbench/api/browser/mainThreadNotebookDocumentsAndEditors.ts:370]: Copied this for the most part from the query editor code.

**Review by Charles-Gagnon — APPROVED** [review]
🎉 Yay!

**Comment by corivera** [comment]
Currently working out some issues with the naming. Sometimes the numbers skip far ahead when opening new untitled notebooks after closing others, so some closed notebooks might not be getting cleaned up properly.

**Pull request merged**